### PR TITLE
PP 172 Error handling for gateway timeouts

### PR DESCRIPTION
**Pull request by acceleanu**
- add error handling for gateway timeouts
- refactoring to use wiremock
- bug fix to make the end to end tests pass

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -6,8 +6,9 @@
     <artifactId>pay-connector</artifactId>
 
     <properties>
-        <dropwizard.version>0.8.1</dropwizard.version>
+        <dropwizard.version>0.8.4</dropwizard.version>
         <mockserver.version>3.9.17</mockserver.version>
```
Comment (zadsas) [inline pom.xml]: Remove this property if we do not use the dependency anymore.

File: src/main/java/uk/gov/pay/connector/app/ConnectorApp.java [hunk 2]
```diff
@@ -72,12 +76,13 @@ public void run(ConnectorConfiguration conf, Environment environment) throws Exc
...
-                                conf.getSmartpayConfig().getNotification()),
-                        "",
-                        String.class)));
+        environment.jersey().register(
+                AuthFactory.binder(
+                        new BasicAuthFactory<>(
+                                new SmartpayAuthenticator(
+                                        conf.getSmartpayConfig().getNotification()),
+                                "",
+                                String.class)));
```
Comment (zadsas) [inline src/main/java/uk/gov/pay/connector/app/ConnectorApp.java]: These are just formatting changes. Please revert them.

Comment (acceleanu) [inline src/main/java/uk/gov/pay/connector/app/ConnectorApp.java]: I actually find this more readable.

File: pom.xml (hunk 1 above)
Comment (acceleanu) [inline pom.xml]: removed

File: src/main/java/uk/gov/pay/connector/service/CardService.java [hunk 3]
```diff
@@ -103,7 +103,8 @@ private GatewayResponse captureFor(String chargeId, Map<String, Object> charge)
     private GatewayResponse authoriseFor(String chargeId, Card cardDetails, Map<String, Object> charge) {
 
         AuthorisationRequest request = authorisationRequest(String.valueOf(charge.get(CHARGE_ID_KEY)), String.valueOf(charge.get(AMOUNT_KEY)), cardDetails);
-        AuthorisationResponse response = paymentProviderFor(charge).authorise(request);
+        AuthorisationResponse response = paymentProviderFor(charge)
+                .authorise(request);
```
Comment (zadsas) [inline src/main/java/uk/gov/pay/connector/service/CardService.java]: Could you please revert this change that does not add any value?

Comment (acceleanu) [inline src/main/java/uk/gov/pay/connector/service/CardService.java]: Again, I find it more clear.

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubITest.java [hunk 4]
```diff
@@ -1,56 +1,78 @@
...
-    public MockServerRule smartpayMockRule = new MockServerRule(this);
+    public WireMockRule wireMockRule = new WireMockRule(port);
 
     @Rule
     public DropwizardAppWithPostgresRule app = new DropwizardAppWithPostgresRule(
-            config("smartpay.url", "http://localhost:" + smartpayMockRule.getHttpPort())
+            config("smartpay.url", "http://localhost:" + port + "/pal/servlet/soap/Payment")
     );
 
     private SmartpayMockClient smartpayMock;
```
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubITest.java]: As I understand it, this is not a client. Maybe something like SmartpayStub?

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubInvalidUrlITest.java [hunk 5]
```diff
@@ -45,10 +46,7 @@ public void setup() {
     public void failedCapture_InvalidConnectorUrl() throws Exception {
```
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubInvalidUrlITest.java]: We would need another test for authorisation.

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketErrorITest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    private DatabaseTestHelper db;
+
+    @Before
+    public void setup() {
+        db = app.getDatabaseTestHelper();
+    }
+
+    @Test
+    public void failedCapture_SocketErrorFromGateway() throws Exception {
```
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketErrorITest.java]: We would need another test for authorisation.

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketReadTimeoutITest.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    private DatabaseTestHelper db;
+
+    @Before
+    public void setup() {
+        smartpayMock = new SmartpayMockClient(TRANSACTION_ID);
+        db = app.getDatabaseTestHelper();
+    }
+
+    @Test
+    public void failedCapture_ConnectionTimeoutFromGateway() throws Exception {
```
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketReadTimeoutITest.java]: We would need another test for authorisation.

File: src/test/java/uk/gov/pay/connector/rules/WorldpayMockClient.java [hunk 8]
```diff
@@ -1,23 +1,18 @@
...
 
-import static javax.ws.rs.HttpMethod.POST;
+import static com.github.tomakehurst.wiremock.client.WireMock.*;
+import static javax.ws.rs.core.HttpHeaders.CONTENT_TYPE;
 import static javax.ws.rs.core.MediaType.TEXT_XML;
-import static org.mockserver.model.HttpRequest.request;
-import static org.mockserver.model.HttpResponse.response;
 import static uk.gov.pay.connector.util.TransactionId.randomId;
 
 public class WorldpayMockClient {
```
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/rules/WorldpayMockClient.java]: As I understand it, this is not a client. Maybe something like WorldpayStub?

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketErrorITest.java (hunk 6 above)
Comment (acceleanu) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketErrorITest.java]: - with the move of error handling of response codes to the GatewayClient it will test the same code path

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketReadTimeoutITest.java (hunk 7 above)
Comment (acceleanu) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketReadTimeoutITest.java]: - with the move of error handling of response codes to the GatewayClient it will test the same code path

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketErrorITest.java (hunk 6 above)
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketErrorITest.java]: :+1:

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketReadTimeoutITest.java (hunk 7 above)
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubSocketReadTimeoutITest.java]: :+1:

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubInvalidUrlITest.java (hunk 5 above)
Comment (acceleanu) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubInvalidUrlITest.java]: - with the move of error handling of response codes to the GatewayClient it will test the same code path

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubITest.java (hunk 4 above)
Comment (acceleanu) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubITest.java]: renamed

File: src/test/java/uk/gov/pay/connector/it/client/SmartpayStubInvalidUrlITest.java (hunk 5 above)
Comment (zadsas) [inline src/test/java/uk/gov/pay/connector/it/client/SmartpayStubInvalidUrlITest.java]: :+1:
